File: native/windows/SkillFlow/SkillFlow.Tests/DaemonClientTests.cs
using System.Net;
using System.Text.Json;
using SkillFlow.Daemon;
using Xunit;

namespace SkillFlow.Tests;

public sealed class RecordingHandler : HttpMessageHandler
{
    private readonly Func<HttpRequestMessage, HttpResponseMessage> _handler;

    public RecordingHandler(Func<HttpRequestMessage, HttpResponseMessage> handler)
    {
        _handler = handler;
    }

    public HttpRequestMessage? Request { get; private set; }

    public string? RequestBody { get; private set; }

    protected override async Task<HttpResponseMessage> SendAsync(
        HttpRequestMessage request,
        CancellationToken cancellationToken)
    {
        Request = request;
        RequestBody = request.Content is null
            ? null
            : await request.Content.ReadAsStringAsync(cancellationToken);
        return _handler(request);
    }
}

public sealed class DaemonClientTests
{
    [Fact]
    public async Task InvokeAsyncSendsTokenVersionAndNativeEnvelope()
    {
        var statePath = await WriteEndpointAsync();
        var handler = new RecordingHandler(_ => JsonResponse("""
        {"ok":true,"result":{"ok":true,"result":[{"name":"Codex Skill"}],"error":null}}
        """));
        using var httpClient = new HttpClient(handler);
        var client = new DaemonClient(httpClient, statePath);

        var skills = await client.InvokeAsync<List<SkillFixture>>(
            "skills.list",
            requestId: "request-001");

        Assert.Equal(["Codex Skill"], skills.Select(skill => skill.Name));
        Assert.Equal("http://127.0.0.1:49151/invoke", handler.Request?.RequestUri?.ToString());
        Assert.Equal("POST", handler.Request?.Method.Method);
        Assert.Equal("test-token", handler.Request?.Headers.GetValues("X-SkillFlow-Token").Single());

        using var document = JsonDocument.Parse(handler.RequestBody ?? string.Empty);
        var root = document.RootElement;
        var parameters = root.GetProperty("params");

        Assert.Equal("native.api", root.GetProperty("method").GetString());
        Assert.Equal(DaemonClient.ApiVersion, parameters.GetProperty("version").GetString());
        Assert.Equal("skills.list", parameters.GetProperty("method").GetString());
        Assert.Equal("request-001", parameters.GetProperty("requestID").GetString());
        Assert.Equal(JsonValueKind.Null, parameters.GetProperty("params").ValueKind);
    }

    [Fact]
    public async Task InvokeAsyncMapsMethodNotFoundError()
    {
        var statePath = await WriteEndpointAsync();
        var handler = new RecordingHandler(_ => JsonResponse("""
        {"ok":true,"result":{"ok":false,"result":null,"error":{"code":"method_not_found","message":"method not found","messageKey":"nativeapi.error.method_not_found"}}}
        """));
        using var httpClient = new HttpClient(handler);
        var client = new DaemonClient(httpClient, statePath);

        var exception = await Assert.ThrowsAsync<DaemonClientException>(() =>
            client.InvokeAsync<List<SkillFixture>>("missing.method"));

        Assert.Equal(DaemonClientErrorKind.Api, exception.Kind);
        Assert.Equal("method_not_found", exception.Code);
        Assert.Equal("nativeapi.error.method_not_found", exception.MessageKey);
    }

    [Fact]
    public async Task InvokeAsyncMapsUnauthorizedResponse()
    {
        var statePath = await WriteEndpointAsync();
        var handler = new RecordingHandler(_ => new HttpResponseMessage(HttpStatusCode.Unauthorized));
        using var httpClient = new HttpClient(handler);
        var client = new DaemonClient(httpClient, statePath);

        var exception = await Assert.ThrowsAsync<DaemonClientException>(() =>
            client.InvokeAsync<List<SkillFixture>>("skills.list"));

        Assert.Equal(DaemonClientErrorKind.Unauthorized, exception.Kind);
    }

    [Fact]
    public async Task InvokeAsyncMapsMissingEndpointFile()
    {
        var statePath = Path.Combine(
            Path.GetTempPath(),
            $"skillflow-missing-{Guid.NewGuid():N}.json");
        using var httpClient = new HttpClient(new RecordingHandler(_ => new HttpResponseMessage()));
        var client = new DaemonClient(httpClient, statePath);

        var exception = await Assert.ThrowsAsync<DaemonClientException>(() =>
            client.InvokeAsync<List<SkillFixture>>("skills.list"));

        Assert.Equal(DaemonClientErrorKind.EndpointUnavailable, exception.Kind);
    }

    private static HttpResponseMessage JsonResponse(string json) => new(HttpStatusCode.OK)
    {
        Content = new StringContent(json, System.Text.Encoding.UTF8, "application/json")
    };

    private static async Task<string> WriteEndpointAsync()
    {
        var path = Path.Combine(
            Path.GetTempPath(),
            $"skillflow-daemon-{Guid.NewGuid():N}.json");
        await File.WriteAllTextAsync(path, """
        {"address":"127.0.0.1:49151","token":"test-token","pid":12345}
        """);
        return path;
    }

    private sealed record SkillFixture(string Name);
}

File: native/macos/SkillFlow/Tests/DaemonClientTests.swift
import Foundation
import XCTest

#if canImport(SkillFlowCore)
@testable import SkillFlowCore
#else
@testable import SkillFlow
#endif

final class MockURLProtocol: URLProtocol {
    static var handler: ((URLRequest) throws -> (HTTPURLResponse, Data))?

    override class func canInit(with request: URLRequest) -> Bool { true }
    override class func canonicalRequest(for request: URLRequest) -> URLRequest { request }

    override func startLoading() {
        guard let handler = MockURLProtocol.handler else {
            client?.urlProtocol(self, didFailWithError: URLError(.unsupportedURL))
            return
        }
        do {
            let (response, data) = try handler(request)
            client?.urlProtocol(self, didReceive: response, cacheStoragePolicy: .notAllowed)
            client?.urlProtocol(self, didLoad: data)
            client?.urlProtocolDidFinishLoading(self)
        } catch {
            client?.urlProtocol(self, didFailWithError: error)
        }
    }

    override func stopLoading() {}
}

final class DaemonClientTests: XCTestCase {
    override func tearDown() {
        MockURLProtocol.handler = nil
        super.tearDown()
    }

    func testInvokeSendsTokenVersionAndNativeEnvelope() async throws {
        let stateURL = try writeEndpoint(address: "127.0.0.1:49151", token: "test-token")
        let client = DaemonClient(endpointPath: stateURL.path, session: mockedSession())
        let captured = LockedBox<URLRequest?>(nil)

        MockURLProtocol.handler = { request in
            captured.value = request
            let body = """
            {"ok":true,"result":{"ok":true,"result":[{"name":"Codex Skill"}],"error":null}}
            """
            return (
                Self.httpResponse(status: 200),
                Data(body.utf8)
            )
        }

        let skills: [SkillFixture] = try await client.invoke(
            "skills.list",
            requestID: "request-001"
        )

        XCTAssertEqual(skills.map(\.name), ["Codex Skill"])
        let request = try XCTUnwrap(captured.value)
        XCTAssertEqual(request.url?.absoluteString, "http://127.0.0.1:49151/invoke")
        XCTAssertEqual(request.value(forHTTPHeaderField: "X-SkillFlow-Token"), "test-token")
        XCTAssertEqual(request.httpMethod, "POST")

        let body = try XCTUnwrap(request.bodyData)
        let envelope = try XCTUnwrap(
            JSONSerialization.jsonObject(with: body) as? [String: Any]
        )
        XCTAssertEqual(envelope["method"] as? String, "native.api")

        let parameters = try XCTUnwrap(envelope["params"] as? [String: Any])
        XCTAssertEqual(parameters["version"] as? String, NativeAPIVersion.current)
        XCTAssertEqual(parameters["method"] as? String, "skills.list")
        XCTAssertEqual(parameters["requestID"] as? String, "request-001")
        XCTAssertTrue(parameters["params"] is NSNull)
    }

    func testInvokeMapsMethodNotFoundError() async throws {
        let stateURL = try writeEndpoint(address: "127.0.0.1:49151", token: "test-token")
        let client = DaemonClient(endpointPath: stateURL.path, session: mockedSession())

        MockURLProtocol.handler = { _ in
            let body = """
            {"ok":true,"result":{"ok":false,"result":null,"error":{"code":"method_not_found","message":"method not found","messageKey":"nativeapi.error.method_not_found"}}}
            """
            return (
                Self.httpResponse(status: 200),
                Data(body.utf8)
            )
        }

        do {
            let _: [SkillFixture] = try await client.invoke("missing.method")
            XCTFail("Expected method_not_found to be thrown")
        } catch let error as DaemonClientError {
            XCTAssertEqual(
                error,
                .api(
                    code: "method_not_found",
                    messageKey: "nativeapi.error.method_not_found",
                    message: "method not found"
                )
            )
        }
    }

    func testInvokeMapsUnauthorizedResponse() async throws {
        let stateURL = try writeEndpoint(address: "127.0.0.1:49151", token: "expired-token")
        let client = DaemonClient(endpointPath: stateURL.path, session: mockedSession())

        MockURLProtocol.handler = { _ in
            (
                Self.httpResponse(status: 401),
                Data(#"{"ok":false,"error":"unauthorized"}"#.utf8)
            )
        }

        do {
            let _: [SkillFixture] = try await client.invoke("skills.list")
            XCTFail("Expected unauthorized to be thrown")
        } catch let error as DaemonClientError {
            XCTAssertEqual(error, .unauthorized)
        }
    }

    func testMissingEndpointFileMapsToEndpointUnavailable() async {
        let missingPath = NSTemporaryDirectory() + "/skillflow-missing-\(UUID().uuidString).json"
        let client = DaemonClient(endpointPath: missingPath, session: mockedSession())

        do {
            let _: [SkillFixture] = try await client.invoke("skills.list")
            XCTFail("Expected endpointUnavailable to be thrown")
        } catch let error as DaemonClientError {
            XCTAssertEqual(error, .endpointUnavailable)
        }
    }
}

private struct SkillFixture: Codable, Equatable {
    let name: String
}


private final class LockedBox<Value> {
    private let lock = NSLock()
    private var storage: Value

    init(_ value: Value) {
        self.storage = value
    }

    var value: Value {
        get {
            lock.lock()
            defer { lock.unlock() }
            return storage
        }
        set {
            lock.lock()
            defer { lock.unlock() }
            storage = newValue
        }
    }
}

private extension DaemonClientTests {
    func mockedSession() -> URLSession {
        let configuration = URLSessionConfiguration.ephemeral
        configuration.protocolClasses = [MockURLProtocol.self]
        return URLSession(configuration: configuration)
    }

    func writeEndpoint(address: String, token: String) throws -> URL {
        let url = URL(fileURLWithPath: NSTemporaryDirectory())
            .appendingPathComponent("skillflow-daemon-\(UUID().uuidString).json")
        let endpoint = """
        {"address":"\(address)","token":"\(token)","pid":12345}
        """
        try Data(endpoint.utf8).write(to: url)
        return url
    }

    static func httpResponse(status: Int) -> HTTPURLResponse {
        HTTPURLResponse(
            url: URL(string: "http://127.0.0.1:49151/invoke")!,
            statusCode: status,
            httpVersion: "HTTP/1.1",
            headerFields: ["Content-Type": "application/json"]
        )!
    }
}

private extension URLRequest {
    var bodyData: Data? {
        if let body = httpBody {
            return body
        }
        guard let stream = httpBodyStream else {
            return nil
        }
        stream.open()
        defer { stream.close() }

        var data = Data()
        let bufferSize = 4096
        let buffer = UnsafeMutablePointer<UInt8>.allocate(capacity: bufferSize)
        defer { buffer.deallocate() }

        while stream.hasBytesAvailable {
            let read = stream.read(buffer, maxLength: bufferSize)
            if read <= 0 {
                break
            }
            data.append(buffer, count: read)
        }
        return data
    }
}
